crucible-llvm/src/Lang/Crucible/LLVM/SimpleLoopFixpointCHC.hs

```diff
@@ -0,0 +1,1415 @@
+                          ["unexpected storage type " ++ show storage_type]
+
+                C.MemArrayStore arr _ -> do
+                  join_variable <- W4.freshConstant sym (W4.safeSymbol "mem_join_var") (W4.exprType arr)
+                  return $ MemArrayFixpointEntry join_variable sz
+
+                _ -> fail "SimpleLoopFixpoint.joinMem: unsupported MemSource"
+
+              return $ Just (mem_loc, mem_entry)
+
+            else
+              return Nothing
+
+        _ -> fail $ "SimpleLoopFixpoint: not MemWrite: " ++ show (C.ppMemWrites mem_writes))
+      (List.concat $ IntMap.elems indexed_writes)
+
+    C.MemWrites [] -> return Map.empty
+
+    _ -> fail $ "SimpleLoopFixpoint: not MemWritesChunkIndexed: " ++ show (C.ppMemWrites mem_writes)
+
+  checkDisjointRegions sym $ Map.keys mem_subst
+
+  return mem_subst
+
+checkDisjointRegions ::
+  (C.IsSymInterface sym, C.HasPtrWidth wptr) =>
+  sym ->
+  [MemLocation sym wptr] ->
+  IO ()
+checkDisjointRegions sym = \case
+  hd_mem_loc : tail_mem_locs -> do
+    mapM_ (checkDisjointRegions' sym hd_mem_loc) tail_mem_locs
+    checkDisjointRegions sym tail_mem_locs
+  [] -> return ()
+
+checkDisjointRegions' ::
+  (C.IsSymInterface sym, C.HasPtrWidth wptr) =>
+  sym ->
+  MemLocation sym wptr ->
+  MemLocation sym wptr ->
+  IO ()
+checkDisjointRegions' sym mem_loc1 mem_loc2 = do
+  ptr1 <- memLocationPtr sym mem_loc1
+  ptr2 <- memLocationPtr sym mem_loc2
+  disjoint_pred <- C.buildDisjointRegionsAssertionWithSub
+    sym
+    ptr1
+    (memLocationSize mem_loc1)
+    ptr2
+    (memLocationSize mem_loc2)
+  when (W4.asConstantPred disjoint_pred /= Just True) $
+    fail $
+      "SimpleLoopFixpoint: non-disjoint ranges: off1="
+      ++ show (W4.printSymExpr $ C.llvmPointerOffset ptr1)
+      ++ ", sz1="
+      ++ show (W4.printSymExpr $ memLocationSize mem_loc1)
+      ++ ", off2="
+      ++ show (W4.printSymExpr $ C.llvmPointerOffset ptr2)
+      ++ ", sz2="
+      ++ show (W4.printSymExpr $ memLocationSize mem_loc2)
+
+
+loadMemJoinVariables ::
+  (C.IsSymBackend sym bak, C.HasPtrWidth wptr, C.HasLLVMAnn sym, ?memOpts :: C.MemOptions) =>
+  bak ->
+  C.MemImpl sym ->
+  Map (MemLocation sym wptr) (MemFixpointEntry sym wptr) ->
+  IO (MapF (W4.SymExpr sym) (W4.SymExpr sym))
+loadMemJoinVariables bak mem subst =
+  let sym = C.backendGetSym bak in
+  MapF.fromList . catMaybes <$> mapM
+    (\(mem_loc, mem_entry) -> do
+      ptr <- memLocationPtr sym mem_loc
+      case mem_entry of
+        MemStoreFixpointEntry join_variable storage_type -> do
+          val <- C.doLoad bak mem ptr storage_type (C.LLVMPointerRepr $ W4.bvWidth join_variable) C.noAlignment
+          case W4.asNat (C.llvmPointerBlock val) of
+                Just 0 -> return $ Just $  MapF.Pair join_variable $ C.llvmPointerOffset val
+                _ -> fail $ "SimpleLoopFixpoint.loadMemJoinVariables: unexpected val:" ++ show (C.ppPtr val)
+          -- foo <- C.loadRaw sym mem ptr storage_type C.noAlignment
```

Comment:
We should remove this commented-out section of code, or else add a prefix comment indicating what it represents and why it's being kept.

Reply:
Nevermind: I see a lot of commented-out code below.  I suppose we should keep it at the moment as "notes" for future refinement rather than just stripping it all out right now; figuring out whether it contains useful information is a fairly large task and I don't want to hold this PR up too much at this stage.